design/pipeline-resources.md

```diff
@@ -0,0 +1,199 @@
+# Resources in YAML
+
+Any external service that is consumed as part of your pipeline is a resource. 
+
+An example of a resource can be another CI/CD pipeline that produces artifacts (say Azure pipelines, Jenkins etc.), code repositories (GitHub, Azure Repos, Git), container image registries (ACR, Docker hub etc.) or package feeds (Azure artifact feed, Artifactor etc.).  
+
+## Why resources?
+
+Resources are defined at one place and can be consumed anywhere in your pipeline. Resources provide you the full traceablity of the services consumed in your pipeline including the branch, version, tags, associated commits and work-items. You can fully automate your DevOps workflow by subscribing to trigger events on your resources.
+
+Resources in YAML represent sources of types pipelines, repositories, containers and packages.
+
+
+### Schema
+
+```yaml
+resources:
+  pipelines: [ pipeline ]  
+  repositories: [ repository ]
+  containers: [ container ]
+  packages: [ package ]
+```
+
+---
+
+## Resources: `pipelines`
+
+If you have a pipeline that produces artifacts, you can consume the artifacts by defining a `pipelines` resource. A pipeline can be another Azure DevOps pipeline or any external pipelines like Jenkins etc.
+
+
+### Schema
+
+```yaml
+resources:        # types: pipelines | repositories | containers | packages
+  pipelines:
+  - pipeline: string  # identifier for the pipeline resource
+    type: enum  # type of the pipeline source like azurePipelines, Jenkins etc. 
+    connection: string  # service connection to connect to the source
+    source:  
```

Comment:
I get why `source` needs to be its own dictionary now, since Jenkins and Azure Pipelines (and TeamCity, etc.) will probably want different inputs, but it seems a bit kludgy. Could we maybe elevate some of what's in `source` to be top-level keys? Then, maybe use something like `inputs: { string: string }` for additional, source-specific information, so that we're consistent with Tasks? @vtbassmatt what do you think?

Reply:
We looked at the inputs approach as well. Ultimately, both are similar approaches. This approach increases readability. I leave it to @vtbassmatt to suggest the better approach.

Reply:
Discussed with @vtbassmatt and @chrisrpatterson about the schema. The inputs starting from `project` to `tags` are provider `type` specific properties. The above schema is for the type: azurePipelines. The inputs may change for other provider types like Jenkins etc.